platform-core-module/src/test/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionsDeleterTest.java

```diff
@@ -0,0 +1,238 @@
+		testClock.setInstant(Instant.ofEpochMilli(CURRENT_DAYTIME.toMillis()));
+	}
+
+	@Test
+	public void testDeleteWithProgramWithZeroRetentionDaysOfExecutions() {
+
+		insertProgramWithZeroRetentionDaysOfExecutionsAndInsertItsExecutions();
+		runProgramExecutionsDeleterAndValidate();
+	}
+
+	@Test
+	public void testDeleteWithProgramWithOneRetentionDaysOfExecutions() {
+
+		insertProgramWithOneRetentionDaysOfExecutionsAndInsertItsExecutions();
+		runProgramExecutionsDeleterAndValidate();
+	}
+
+	@Test
+	public void testDeleteWithProgramWithTwoRetentionDaysOfExecutions() {
+
+		insertProgramWithTwoRetentionDaysOfExecutionsAndInsertItsExecutions();
+		runProgramExecutionsDeleterAndValidate();
+	}
+
+	@Test
+	public void testDeleteWithSeveralExecutionsOnEachDay() {
+
+		AGProgram program = insertProgram(0);
+		expectedRetainedProgramExecutionsMap.addToSet(program, insertExecutionOfProgram(program, ExecutionDay.TODAY));
+		expectedRetainedProgramExecutionsMap.addToSet(program, insertExecutionOfProgram(program, ExecutionDay.TODAY));
+		insertExecutionOfProgram(program, ExecutionDay.YESTERDAY);
+		insertExecutionOfProgram(program, ExecutionDay.YESTERDAY);
+		insertExecutionOfProgram(program, ExecutionDay.TWO_DAYS_AGO);
+		insertExecutionOfProgram(program, ExecutionDay.TWO_DAYS_AGO);
+
+		runProgramExecutionsDeleterAndValidate();
+	}
+
+	@Test
+	public void testDeleteWithNotTerminatedExecution() {
+
+		AGProgram program = insertProgram(0);
+		expectedRetainedProgramExecutionsMap.addToSet(program, insertProgramExecutionWithLog(program, null));
+
+		runProgramExecutionsDeleterAndValidate();
+	}
+
+	@Test
+	public void testDeleteWithSeveralProgramsAndExecutions() {
+
+		insertProgramWithZeroRetentionDaysOfExecutionsAndInsertItsExecutions();
+		insertProgramWithOneRetentionDaysOfExecutionsAndInsertItsExecutions();
+		insertProgramWithTwoRetentionDaysOfExecutionsAndInsertItsExecutions();
+
+		runProgramExecutionsDeleterAndValidate();
+	}
+
+	@Test
+	public void testDeleteWithExecutionTerminatedJustBeforeMidnight() {
+
+		DayTime yesterdayJustBeforeMidnight = ExecutionDay.TODAY.dayTime.minusSeconds(1.0);
+
+		AGProgram program = insertProgram(0);
+		insertProgramExecutionWithLog(program, yesterdayJustBeforeMidnight);
+
+		runProgramExecutionsDeleterAndValidate();
+	}
+
+	private void runProgramExecutionsDeleterAndValidate() {
+
+		new ProgramExecutionsDeleter(0).delete();
+		assertEquals(expectedRetainedProgramExecutionsMap, loadProgramExecutions());
+	}
+
+	private SetMap<AGProgram, AGProgramExecution> loadProgramExecutions() {
+
+		SetMap<AGProgram, AGProgramExecution> programExecutions = new SetMap<>();
+		AGProgramExecution.createSelect().forEach(record -> programExecutions.addToSet(AGProgram.loadOrInsert(record.getProgramUuid()), record));
+		return programExecutions;
+	}
+
+	private void insertProgramWithZeroRetentionDaysOfExecutionsAndInsertItsExecutions() {
```

Comment:
Those `insertProgramWith*` methods basically perform the test setup _and_ define expectations for the assertions. We usually separate those aspects/concerns.
This _can_ surely be done this way. I just find it difficult to wrap my head around this.

Reply:
well, I think, doing setup and defining expected values at the same time saves us many code lines

Reply:
I think unit tests are all about clarity and explicitness, and not so much about low redundancy. In that regard, they're quite different from regular code.
My point is: There's only one single `assertEquals` method in that test class. It compares two maps.
If this assertion ever fails, the cause is _not_ obvious. All we know in this case is, that an "assertion failed" (the equals-based comparison of two maps). We would then need to start _debugging the unit test_ (!) in order to identify the actual problem.

Reply:
Okay, I changed the test class.